Change calibrate field coordinates to action

jmblixt3 opened the pull request:
Field coordinates was an async python function
To make code more ros like it was changed to an
action server/client
It is contained in the same python executor due to
rclpy:1123
So any python action server/client will have to be in the same executor

Isopod00 commented on src/rovr_control/rovr_control/main_control_node.py:
```diff
@@ -117,37 +122,32 @@ def __init__(self) -> None:
-            self.get_logger().warn("Waiting for the lift/zero service to be available (BLOCKING)")
-        self.cli_lift_zero.call_async(Stop.Request())  # Zero the lift by slowly raising it up
+        # while not self.cli_lift_zero.wait_for_service(timeout_sec=1):
+        #     self.get_logger().warn("Waiting for the lift/zero service to be available (BLOCKING)")
```
Uncomment this before we merge this PR into main, we don't want this code commented out in main.

jmblixt3 replied:
fixed

Isopod00 commented on src/rovr_control/rovr_control/main_control_node.py:
```diff
@@ -387,16 +355,22 @@ def joystick_callback(self, msg: Joy) -> None:
+            if self.field_calibrated.done():
+                self.get_logger().info("Field calibration button pressed")
+                if not self.act_calibrate_field_coordinates.wait_for_server(timeout_sec=1.0):
+                    self.get_logger().error("Field calibration action not available")
```
I can test this on my desktop PC when I get a chance later this week

Isopod00 commented on src/rovr_control/rovr_control/calibrate_field_coordinate_server.py:
```diff
@@ -0,0 +1,98 @@
+            self.get_logger().error("Nav2 service not available")
+            return result
+
+        for i in range(1, 3):
```
In the original calibration method, the robot would keep turning in place until an Apriltag is seen. This whole for loop approach seems to change that. At first glance, this for loop seemed strange to me, but I guess it does have the benefit of failing at some point if there is legitimately no Apriltag, whereas the old method would have just kept spinning forever. So maybe this way is better? It does look a bit weird but that might just be me. I can test this new method on my desktop PC when I get a chance later in the week.

jmblixt3 replied:
This just does 2 180 degree turns then fails if none is found

Isopod00 replied:
Doesn't it turn by 180 degrees in the first iteration and 360 degrees in the second iteration? Since your spin goal is i * math.pi

jmblixt3 replied:
I thought it turned to a heading,but if its distance then this is incorrect

Isopod00 replied:
Yeah, I think it's distance

Isopod00 replied:
Okay, I just fixed this

Isopod00 reviewed (changes requested):
I can't test this PR right now but when I have access to my desktop PC later in the week I can.

Isopod00 commented on src/rovr_control/rovr_control/main_control_node.py:
```diff
@@ -387,16 +353,22 @@ def joystick_callback(self, msg: Joy) -> None:
-                self.field_calibrated = False
+            if self.field_calibrated.done():
+                self.get_logger().info("Field calibration button pressed")
+                if not self.act_calibrate_field_coordinates.wait_for_server(timeout_sec=1.0):
```
self.act_calibrate_field_coordinates does not seem to be defined anywhere, which causes the node to crash

Isopod00 reviewed (changes requested):
The current state of this PR does not work, this PR needs to be fixed before merging.

jmblixt3 committed:
Change calibrate field coordinates to action
Field coordinates was an async python function
To make code more ros like it was changed to an
action server/client
It is contained in the same python file due to
rclpy:1123
So any python action server/client will have to be in the same file

Signed-off-by: Jonathan Blixt <jmblixt3@gmail.com>

Isopod00 reviewed (approved):
It works now, looks good to me!
